storage/dataflux/worksteal.go

```diff
@@ -45,16 +56,171 @@ type worker struct {
+		// Worker is active when it finds work in range channel.
+		if w.status == idle {
+			if len(w.lister.ranges) == 0 {
+				time.Sleep(time.Millisecond * time.Duration(sleepDurationWhenIdle))
+				continue
+			} else {
+				newRange := <-w.lister.ranges
+				<-w.idleChannel
+				w.updateWorker(newRange.startRange, newRange.endRange, active)
+			}
+		}
+		// Active worker to list next page of objects within the range.
+		nextPageResult, err := objectLister(ctx, nextPageOpts{
+			startRange:           w.startRange,
+			endRange:             w.endRange,
+			bucketHandle:         w.lister.bucket,
+			query:                w.lister.query,
+			skipDirectoryObjects: w.lister.skipDirectoryObjects,
+			generation:           w.generation,
+		})
+		if err != nil {
+			return fmt.Errorf("listing next page for worker ID %v,  err: %w", w.goroutineID, err)
+		}
+
+		// Append objects listed by objectLister to result.
+		w.result.mu.Lock()
+		w.result.objects = append(w.result.objects, nextPageResult.items...)
+		w.result.mu.Unlock()
+
+		// Listing completed for default page size for the given range.
+		// Update current worker start range to new range and generation
+		// of the last objects listed if versions is true.
+		w.startRange = nextPageResult.nextStartRange
+		w.generation = nextPageResult.generation
+
+		// If listing is complete for the range, make worker idle and continue.
+		if nextPageResult.doneListing {
+			w.status = idle
+			w.idleChannel <- 1
+			w.generation = int64(0)
+			continue
+		}
+		// If listing not complete and idle workers are available, split the range and give half of work to idle worker.
+		if len(w.idleChannel)-len(w.lister.ranges) > 0 && ctx.Err() == nil {
+			// Split range and upload half of work for idle worker.
+			splitPoint, err := w.rangesplitter.splitRange(w.startRange, w.endRange, 1)
+			if err != nil {
+				return fmt.Errorf("splitting range for worker ID:%v, err: %w", w.goroutineID, err)
+			}
+			// If split point is empty, skip splitting the work.
+			if len(splitPoint) < 1 {
+				continue
+			}
+			w.lister.ranges <- &listRange{startRange: splitPoint[0], endRange: w.endRange}
+
+			// Update current worker range.
+			w.endRange = splitPoint[0]
+		}
+	}
 	return nil
 }
 
-// newObjectListerOpts specifies options for instantiating the NewObjectLister.
-type newObjectListerOpts struct {
+// shutDownSignal returns true if all the workers are idle and the or number of objects listed is equal to page size.
+func (w *worker) shutDownSignal() bool {
+	// If all the workers are idle and range channel is empty, no more objects to list.
+	if len(w.idleChannel) == w.lister.parallelism && len(w.lister.ranges) == 0 {
+		return true
+	}
+	// If number of objects listed is equal to the given batchSize, then shutdown.
+	// If batch size is not given i.e. 0, then list until all objects have been listed.
+	if w.lister.batchSize > 0 && len(w.result.objects) >= w.lister.batchSize {
+		return true
+	}
+	return false
```

Comment:
Suggest structuring this like the following (feel free to modify names etc):

```suggestion
	noMoreObjects := len(w.idleChannel) == w.lister.parallelism && len(w.lister.ranges) == 0 
	alreadyListedBatchSizeObjects := w.lister.batchSize > 0 && len(w.result.objects) >= w.lister.batchSize 
	return noMoreObjects || alreadyListedBatchSizeObjects
```